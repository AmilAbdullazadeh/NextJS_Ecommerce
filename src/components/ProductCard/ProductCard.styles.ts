import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles(({ palette, spacing, breakpoints }) => ({
    cardContainer: {
        [breakpoints.down('md')]: {
            padding: spacing(0.25),
        },
        padding: spacing(2),
    },
    cardRoot: {
        width: spacing(60),
        [breakpoints.down('md')]: {
            width: spacing(40),
            margin: spacing(0.25),
            padding: spacing(1),
        },
    },
    productName: {
        display: 'inline',
        marginRight: spacing(1),
    },
    description: {
        height: spacing(10),
        color: palette.text.primary,
        textDecoration: 'none',
        fontSize: 14,
        display: '-webkit-box',
        '-webkit-line-clamp': 4,
        '-webkit-box-orient': 'vertical',
        overflow: 'hidden',
    },
    image: {
        [breakpoints.up('md')]: {
            maxWidth: '50%',
        },
    },
}));

export default useStyles;
