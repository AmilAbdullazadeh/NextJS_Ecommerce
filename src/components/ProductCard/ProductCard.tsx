import React from 'react';
import Card from '@material-ui/core/Card';
import Box from '@material-ui/core/Box';
import Grid from '@material-ui/core/Grid';
import Link from 'next/link';

import { productDetailsPage } from '@core';
import { ImageWithFallback, ProductRating, Typography } from '@components';

import useStyles from './ProductCard.styles';

export interface ProductCardProps {
    id: string;
    name: string;
    ratingValue: number;
    reviewCount: number;
    price: number;
    availability: number;
    description: string;
    image?: string;
    imageFallback: string;
}

const ProductCard: React.FC<ProductCardProps> = ({
    id,
    name,
    ratingValue,
    reviewCount,
    price,
    availability,
    description,
    image,
    imageFallback,
}: ProductCardProps) => {
    const classes = useStyles();

    return (
        <article className={classes.cardContainer}>
            <Card key={id} className={classes.cardRoot}>
                <Grid
                    container
                    direction="row"
                    justifyContent="center"
                    alignItems="center"
                    spacing={2}
                >
                    <Grid item xs={5} md={12}>
                        <Box
                            component="div"
                            display="block"
                            className={classes.image}
                            mx="auto"
                        >
                            <ImageWithFallback
                                src={image ?? ''}
                                fallbackSrc={imageFallback}
                                alt={name}
                                width={200}
                                height={200}
                            />
                        </Box>
                    </Grid>
                    <Grid item xs={7} md={12}>
                        <Box px={1}>
                            <Box component="div">
                                <Link href={productDetailsPage(id)}>
                                    <a className={classes.description}>
                                        <Box
                                            component="h4"
                                            display="inline"
                                            mr={0.5}
                                        >
                                            {name}
                                        </Box>
                                        <Box component="p" display="inline">
                                            {description}
                                        </Box>
                                    </a>
                                </Link>
                            </Box>
                            <ProductRating
                                reviewCount={reviewCount}
                                ratingValue={ratingValue}
                            />
                            <Box
                                display="flex"
                                flexDirection="row"
                                justifyContent="flex-start"
                                mb={2}
                            >
                                <Typography variant="body1" component="span">
                                    {price}â‚¬
                                </Typography>
                                <Box display="inline" ml={1}>
                                    <Typography
                                        variant="body2"
                                        color={
                                            availability ? 'primary' : 'error'
                                        }
                                        component="span"
                                    >
                                        {availability
                                            ? 'in stock'
                                            : 'out of stock'}
                                    </Typography>
                                </Box>
                            </Box>
                        </Box>
                    </Grid>
                </Grid>
            </Card>
        </article>
    );
};

export default ProductCard;
