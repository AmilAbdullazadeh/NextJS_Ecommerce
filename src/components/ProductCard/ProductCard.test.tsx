import React from 'react';

import { render } from '@testUtils';

import ProductCard, { ProductCardProps } from './ProductCard';

describe('ProductCard', () => {
    const defaultProps: ProductCardProps = {
        id: 'mockId',
        name: 'mockName',
        ratingValue: 3.5,
        reviewCount: 10,
        price: 100,
        availability: 200,
        description:
            'Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam',
        imageFallback: 'imageFallbackSrc',
    };

    it('matches snapshot', () => {
        const { container } = render(<ProductCard {...defaultProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('shows in stock message if availability is greater than zero', () => {
        const { getByText } = render(<ProductCard {...defaultProps} />);

        expect(getByText('in stock')).toBeInTheDocument();
    });

    it('shows out of stock message if availability is zero', () => {
        const props: ProductCardProps = { ...defaultProps, availability: 0 };

        const { getByText } = render(<ProductCard {...props} />);

        expect(getByText('out of stock')).toBeInTheDocument();
    });
});
