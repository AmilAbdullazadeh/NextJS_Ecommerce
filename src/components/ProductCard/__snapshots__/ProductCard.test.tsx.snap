// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard matches snapshot 1`] = `
<article
  class="makeStyles-cardContainer-1"
>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-cardRoot-2 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-md-12"
      >
        <div
          class="MuiBox-root MuiBox-root-6 makeStyles-image-5"
        >
          <div
            style="display: block; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="display: block; box-sizing: border-box; padding-top: 100%;"
            />
            <noscript />
            <img
              alt="mockName"
              data-nimg="true"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 MuiGrid-grid-md-12"
      >
        <div
          class="MuiBox-root MuiBox-root-7"
        >
          <div
            class="MuiBox-root MuiBox-root-8"
          >
            <a
              class="makeStyles-description-4"
              href="/products/mockId"
            >
              <h4
                class="MuiBox-root MuiBox-root-9"
              >
                mockName
              </h4>
              <p
                class="MuiBox-root MuiBox-root-10"
              >
                Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam
              </p>
            </a>
          </div>
          <div
            class="MuiBox-root MuiBox-root-11"
          >
            <div
              class="MuiBox-root MuiBox-root-12"
            >
              <span
                aria-label="3.5 Stars"
                class="MuiRating-root MuiRating-sizeSmall MuiRating-readOnly"
                role="img"
              >
                <span
                  class="MuiRating-decimal"
                >
                  <span
                    style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconFilled"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span>
                    <span
                      class="MuiRating-icon MuiRating-iconFilled"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
                <span
                  class="MuiRating-decimal"
                >
                  <span
                    style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconFilled"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span>
                    <span
                      class="MuiRating-icon MuiRating-iconFilled"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
                <span
                  class="MuiRating-decimal"
                >
                  <span
                    style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconFilled"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span>
                    <span
                      class="MuiRating-icon MuiRating-iconFilled"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
                <span
                  class="MuiRating-decimal"
                >
                  <span
                    style="width: 50%; overflow: hidden; z-index: 1; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconFilled"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span>
                    <span
                      class="MuiRating-icon MuiRating-iconEmpty"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
                <span
                  class="MuiRating-decimal"
                >
                  <span
                    style="width: 0%; overflow: hidden; z-index: 1; position: absolute;"
                  >
                    <span
                      class="MuiRating-icon MuiRating-iconEmpty"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span>
                    <span
                      class="MuiRating-icon MuiRating-iconEmpty"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                        />
                      </svg>
                    </span>
                  </span>
                </span>
              </span>
            </div>
            <div
              class="MuiBox-root MuiBox-root-13"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 MuiTypography-colorPrimary"
              >
                10
              </p>
            </div>
          </div>
          <div
            class="MuiBox-root MuiBox-root-14"
          >
            <span
              class="MuiTypography-root MuiTypography-body1"
            >
              100
              €
            </span>
            <div
              class="MuiBox-root MuiBox-root-15"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 MuiTypography-colorPrimary"
              >
                in stock
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</article>
`;
